### ISSUE TITLE
OCC crashes with stackoverflow in hasPackedExpression

### ISSUE THREAD

**Issue by chuggafan** [LADSoft/OrangeC#370]
```cpp

#include <tuple>
#include <typeinfo>
#include <cstdio>
int main()
{
    std::tuple<int, int, int> tup = {12, 12313, 132132};
    auto t = std::make_tuple(1, 123, 142312);
    const std::type_info& t1 = typeid(t);
    printf("%s", t1.name());
}

```

This will crash OCC, this code generates a VERY large expression, and will near-infinitely just call down because of it, after already having a massive number of calls, in fact I got all the way from address  
0x08186868 to 0x03a1b4fc with almost everything inbetween being EXPRESSION*, it's untenable and needs to be fixed ASAP, somehow.

**Comment by chuggafan** [comment]
Solved hasPackedExpression, ran into GetSymRef immediately after with the exact same problem .-.

I have to figure out how to properly do this...

**Comment by chuggafan** [comment]
Now I'm onto copy_expression jesus christ these 10 lines of code have exposed 3 stackoverflow problems already.

I am considering calling this commit "Spring cleaning" at this point because christ that's bad.

**Comment by chuggafan** [comment]
Ok, wow is copy_expression a hard problem, because deep copying a binary tree most solutions are recusrive... @LADSoft or @GitMensch do either of you know how or can think of a way to turn copy_expression into an iterative function?

```cpp

EXPRESSION* copy_expression(EXPRESSION* exp)
{
    EXPRESSION* rv = (EXPRESSION*)(EXPRESSION*)Alloc(sizeof(EXPRESSION));
    *rv = *exp;
    if (rv->left)
        rv->left = copy_expression(rv->left);
    if (rv->right)
        rv->right = copy_expression(rv->right);
    return rv;
}

```

That's there.

Also, my GetSymRef is only a temp fix to at least fix the problem I am immediately having on just one expression side, at least limiting the recursive calls drastically here.

```cpp

EXPRESSION* GetSymRef(EXPRESSION* n)
{
    EXPRESSION* rv = NULL;
    std::stack<EXPRESSION*> st;
    while (true)
    {
        switch (n->type)
        {
            case en_labcon:
            case en_global:
            case en_auto:
            case en_absolute:
            case en_pc:
            case en_threadlocal:
                if (st.size() > 0)
                {
                    goto next;
                }
                return n;
            case en_tempref:
                return NULL;
            default:
                if (n->left)
                {
                    st.push(n);
					n = n->left;
                   continue;
                next:
                    rv = n;
					n = st.top();
                    st.pop();
                }
                if (!rv && n->right)
                    rv = GetSymRef(n->right);
                return rv;
        }
    }
}

```

This is what I currently have, I want to change it to actually be fully iterative, but that's an even bigger challenge...

I'll add these into the "fix crashing" pr, I guess.... same topic, just more issues being targeted

**Comment by GitMensch** [comment]
Both code snippets look fine, where do you see the problem?
The "fix crashing PR" doesn't have this code in yet.

**Comment by chuggafan** [comment]
Yhea.. I got back into problems with the other sections, I didn't add getSymRef to my PR and it looks like my code was a total fail. I've found a nice, ITERATIVE method to replace copy_expression however, and I found it in Knuth's Volume 1, specifically page 328 of the second edition... so I'll look at that first.

**Comment by chuggafan** [comment]
So if all goes well with my most recent push to my PR, the newest roadblock is *drumroll please*... the mangler. Not something easy to fix, but the file now crashes at the mangler, my one change to the mangler was to stop __chkstk from going crazy (presumably because the allocated stack of 8192 was overfilled or is too large to work with). Looks like that entire system will have to be re-written to be iterative and it will not be fun to attempt to do such a major change...

**Comment by LADSoft** [comment]
there is a possibility that this is resolved as part of #366.   I know that most of the samples in #366 were causing similar kinds of recursion and this has been addressed at this point...

**Comment by LADSoft** [comment]
I fixed these kinds of issues here: 1238880ef574b93fbe2ba5cddf9a8b09931a616d

**Comment by GitMensch** [comment]
Hm, do you also have a test case for these kinds of issues to add?

**Comment by LADSoft** [comment]
Yeah I can add my tuples/shared pointer tests.   They exhibited all the problems I closed in relation to issue #366.   Will do that over the next couple of days...
